migrate bucardo pg article and update links to code changes (#1249)
aiven/devportal: docs/products/postgresql/howto/migrate-using-bucardo.rst

Migrating to Aiven for PostgreSQL® using Bucardo
===============================================

The preferred approach to migrating a database to Aiven for PostgreSQL®

> is to use `our Open Source migration tool <https://help.aiven.io/en/articles/4358591>`__ . However, if you

Reviewer: ```suggestion
is to use `Aiven's open source migration tool <https://developer.aiven.io/docs/products/postgresql/concepts/aiven-db-migrate.html>`_. However, if you
```

> are running PostgreSQL 9.6 or do not have ``superuser`` access to your

Reviewer: ```suggestion
are running PostgreSQL 9.6 (or earlier) or do not have ``superuser`` access to your
```

database to add replication slots, you can use the open source
`Bucardo <https://bucardo.org>`__ tool to allow replication to Aiven.

**Requirements:**

> -  An Aiven for PostgreSQL database

Author: Any plan

-  Your current database

-  A computer with Bucardo installed

> -  Connectivity between your current database, the Aiven database, and

Reviewer: ```suggestion
-  Connectivity between Bucardo and the source and target databases
```

>    Bucardo

Reviewer: ```suggestion

```

Moving existing data
~~~~~~~~~~~~~~~~~~~~

To move existing data, you can follow the steps below and
`update <https://bucardo.org/Bucardo/operations/onetimecopy>`__ your
``sync`` job to use the ``onetimecopy`` and move existing data across.
You can also use the standard ``pg_dump`` and ``pg_restore`` commands to
fill the Aiven database and use Bucardo for syncing any changes to the

> source database and ensuring it remains up to date.

Reviewer: ```suggestion
source database and ensuring it remains up-to-date.
```

Replicating changes
~~~~~~~~~~~~~~~~~~~

To migrate your data using Bucardo:

#. | Install Bucardo using `the installation
     instructions <https://bucardo.org/Bucardo/installation/>`__ on the
     Bucardo site.

> #. | Install the ``aiven_extras`` extension to your current database.

Reviewer: ```suggestion
#. | Install the ``aiven_extras`` `extension <https://developer.aiven.io/docs/products/postgresql/concepts/dba-tasks-pg.html#aiven-extras-extension>`_ to your current database.
```

   | Bucardo requires the superuser role to set the
     ``session_replication_role`` parameter. Aiven uses the open source
     ``aiven_extras`` extension to allow you to run ``superuser``
     commands as a different user, as direct ``superuser`` access is not
     provided for security reasons.

#. | Open and edit the ``Bucardo.pm`` file with administrator
     privileges.
   | The location of the file can vary according to your operating
     system, but you might find it in
     ``/usr/local/share/perl5/5.32/Bucardo.pm`` , for example.

   #. Scroll down until you see a ``disable_triggers`` function, on line

>       `5324 <https://github.com/bucardo/bucardo/blob/1ff4d32d1924f3437af3fbcc1a50c1a5b21d5f5c/Bucardo.pm#L5324>`__

Author: They are permalinks to that content on the line so they remain static, as Bucardo is essentially one big Perl script, I would prefer to leave the line numbers in to ensure accurate references in future versions

   | For example:

   ::

      bucardo add db srcdb dbhost=0.0.0.0 dbport=5432 dbname=all_your_base dbuser=$DBUSER dbpass=$DBPASS

      bucardo add db destdb dbhost=cg-pg-dev-sandbox.aivencloud.com dbport=21691 dbname=all_your_base dbuser=$DBUSER dbpass=$DBPASS

#. Add the tables that you want to replicate:

   ::

      bucardo add table belong to us herd=$HERD db=srcdb

   .. note:: You can set ``$HERD`` to any name you choose for the herd, which is used to set up the synchronization.

#. Dump and restore the database from your source to your Aiven service:

   ::

      pg_dump --schema-only --no-owner all_your_base > base.sql
      psql "$AIVEN_DB_URL" < base.sql

   | You can restore the source data or provide only the schema,
     depending on the size of your current database.

#. Create the ``dbgroup`` for Bucardo:

   ::

      bucardo add dbgroup src_to_dest srcdb:source destdb:target
      bucardo add sync sync_src_to_dest relgroup=$HERD db=srcdb,destdb
      (sudo) bucardo start
      bucardo status sync_src_to_dest

#. | Start Bucardo and run the ``status`` command. When ``Current state`` is ``Good`` , the data is flowing to your
     Aiven database.

> #. | Log in to the Aiven web console and go to the **Current Queries** tab for your service. This shows you that the ``bucardo`` process is inserting data.

Reviewer: Log in to the `Aiven web console <https://console.aiven.io>`_ and go to the **Current Queries** tab for your Aiven for PostgreSQL service.

Author: You can make these as suggested changes instead of comments that can be directly resolved. Happy to make this change, though

> #. | Log in to the Aiven web console and go to the **Current Queries** tab for your service. This shows you that the ``bucardo`` process is inserting data.

Reviewer: ```suggestion
#. | Log in to the `Aiven web console <https://console.aiven.io>`_ and go to the **Current Queries** tab for your Aiven for PostgreSQL service. This shows you that the ``bucardo`` process is inserting data.
```

#. Once all your data is synchronized, switch the database connection

>    for your applications to the Aiven service URI.

Reviewer: Question to CG: Should we rephrase as "...switch the database connection
   for your applications to Aiven for PostgreSQL"? The concern is that the service URI includes the service credentials so do we need to be explicit about not using the service URI directly?

Author: True, URI is not needed here and maybe more detail creates more confusion. Your suggestion works for me.

>    for your applications to the Aiven service URI.

Reviewer: ```suggestion
   for your applications to Aiven for PostgreSQL.
```
